pbj-core/pbj-compiler/src/main/java/com/hedera/pbj/compiler/impl/Common.java

```diff
@@ -172,14 +174,367 @@ public static String javaPrimitiveToObjectType(String primitiveFieldType) {
+						f.type() == Field.FieldType.SINT64 ||
+						f.type() == Field.FieldType.UINT64) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Long.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.ENUM) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + $fieldName.hashCode();
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.BOOL) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Boolean.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.FLOAT) {
+					generatedCodeSoFar += (
+               		"""
+             		if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Float.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.DOUBLE) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + Double.hashCode($fieldName);
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.STRING) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + $fieldName.hashCode();
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				} else if (f.type() == Field.FieldType.BYTES) {
+					generatedCodeSoFar += (
+               		"""
+					if ($fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + ($fieldName == null ? 0 : $fieldName.hashCode());
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else if (f.parent() == null) { // process sub message
+				    generatedCodeSoFar += (
+               		"""
+					if ($fieldName != null && $fieldName != DEFAULT.$fieldName) {
+						result = 31 * result + $fieldName.hashCode();
+					}
+					""").replace("$fieldName", f.nameCamelFirstLower());
+				}
+				else {
+					throw new RuntimeException("Unexpected field type for getting HashCode - " + f.type().toString());
+				}
+			}
+		}
+		return generatedCodeSoFar.indent(DEFAULT_INDENT * 2);
+	}
+
+	/**
+	 * Get the hashcode codegen for a optional field.
+	 * @param generatedCodeSoFar The string that the codegen is generated into.
+	 * @param f The field for which to generate the hash code.
+	 * @return Updated codegen string.
+	 */
+	@NonNull
+	private static String getOptionalHashCodeGeneration(String generatedCodeSoFar, Field f) {
+		switch (f.messageType()) {
+			case "StringValue" -> generatedCodeSoFar += (
+     			"""
+				if ($fieldName != DEFAULT.$fieldName) {
```

Comment:
Should it be equals() instead?

Reply:
In this particular context `equals` and reference equality should work exactly the same because the default value is singleton. However, for the sake of clean code I'll replace it with `equals`